### PR TITLE
Add support to install EKS Pod Identity Webhook

### PR DESCRIPTION
**Pull request by h3poteto**
Refs #12436 

I add an option to install EKS Pod Identity Webhook. If you specify it in the cluster spec, the webhook will be installed. And it requires cert-manager to manage certificates of the webhook.
For example:
```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  name: playground
spec:
  certManager:
    enabled: true
    managed: true
  podIdentityWebhook:
    enabled: true
# ...
```

### CODE REVIEW

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:147]: I am not sure if we should add an objectSelector to this one. Getting requests for every single pod in the cluster sounds like a potential issue.

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:156]: Is this actually safe? won't this just end up in a running pod that will start misbehaving?

**Review by olemarkus — CHANGES_REQUESTED** [review]

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:59]: Oh, we don't need this permission.

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:156]: I just copied it from the official repository: https://github.com/aws/amazon-eks-pod-identity-webhook/blob/b19c295a269ad9de50b10e3a9cdc9ec1f7d48a19/deploy/mutatingwebhook.yaml#L10
But I agree that it is not good.

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:147]: Hmm... `objectSelector` is nice, but what value should we set?
The [official repository](https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/deploy/mutatingwebhook.yaml) does not have any objectSelector, so it checks all pods in the cluster.
I guess should we allow users to customize the value of `objectSelector`?

Comment (olemarkus) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:147]: It's a matter of documentation, I think. We can say that only pods with the label `kops.k8s.io/inject-aws-pod-identity` will be evaluated. I am not sure performance is a big concern though. Don't see much mention of it elsewhere.

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:88]: We maybe want more than 1 replica?

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:147]: OK, I see.

Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:130]: I update the document.
https://github.com/kubernetes/kops/pull/13176/commits/0144c1f66141c9de197ffd5d021946c6914e14f2

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:79]: The webhook never runs on the control plane, so you can just set this one to 2.

**Review by olemarkus — COMMENTED** [review]
/ok-to-test

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (olemarkus) [inline upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:1]: this file should be named similar to the other addons

Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:1]: Is it a name like `k8s-1.23.yaml.template`, for example?

Comment (olemarkus) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template:1]: Yeah. Where the version is the minimum version supporting the manifests

File: upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template
Comment (h3poteto) [reply upup/models/cloudup/resources/addons/eks-pod-identity-webhook.addons.k8s.io/v0.3.0.yaml.template]: Removed: https://github.com/kubernetes/kops/pull/13176/commits/3cd20cb82a7bfbe3bf1847c3b158c27ce4e86590

**Comment by olemarkus** [comment]
FYI: I am working on a feature where kops will build a configmap that the EKS hook can read SA to role mapping from. I'd like that feature to go in before we merge this.

**Comment by h3poteto** [comment]
OK, I found your pull request in the amazon-eks-pod-identity-webhook repository.

**Comment by olemarkus** [comment]
@h3poteto sorry inadvertently forced pushed to your branch.

https://github.com/aws/amazon-eks-pod-identity-webhook/pull/142 is taking a while, but I think we can still finish this PR. The implementation may change in that PR, but I expect the functionality/interface will be the same.

I also removed the object selector. The EKS team had not experienced any performance issues, so we do not need to introduce that burden, I think.

**Comment by h3poteto** [comment]
OK, thank you.
I also removed unnecessary descriptions regarding the object selector in the doc.

**Comment by olemarkus** [comment]
Are you able to test the new webhook functionality?

**Comment by h3poteto** [comment]
Hmm...it does not work fine.
Because we specify `failurePolicy: Fail` and remove the object selector. So some required pods can not launch.
```
Feb 20 04:36:47 ip-172-20-37-9 kubelet[5300]: E0220 04:36:47.392591    5300 kubelet.go:1711] "Failed creating a mirror pod for" err="Internal error occurred: failed calling webhook \"pod-identity-webhook.amazonaws.com\": failed to call webhook: Post \"https://pod-identity-webhook.kube-system.svc:443/mutate?timeout=10s\": dial tcp 100.64.59.214:443: connect: connection refused" pod="kube-system/kube-controller-manager-ip-172-20-37-9.ap-northeast-1.compute.internal"
```

**Comment by olemarkus** [comment]
Isn't that error transient? KCM is actually running. The wevhook cannot affect the mirror pod anyway.

**Comment by h3poteto** [comment]
And the Issuer is not created even though I set `NeedsPKI: true`. :confused:

**Comment by olemarkus** [comment]
There is a fun chicken/egg thing happening here:
* pod identity webhook pods won't start because it fails the webhook
* webhook fails because no pod identity webhook pods are running

We need to label pods managed by kops so that it doesn't match the webhook, I think.

**Comment by h3poteto** [comment]
@olemarkus I fixed the object selector to ignore myself and the cert-manager pod. It works fine.
https://github.com/kubernetes/kops/pull/13176/commits/6e56fcc89081d7c2e74d39d2e1a69fb5ffb42a20

**Comment by olemarkus** [comment]
Was that enough? I think perhaps there will be a race here ... say should the webhook be created before dns-controller pod or kops-controller, none of the worker nodes will be able to join.

**Comment by olemarkus** [comment]
A test for the above-mentioned issue would be to do a `kops rolling-update cluster --yes --force --cloudonly`.

I pushed a commit now that will add a label to all pods in all kops addons, then ignore that in the webhook. Also ignoring kube-apiserver as the mirror pod is needed for dns-controller to create the API dns entries.

**Comment by h3poteto** [comment]
Thank you. I agree that the `kops.k8s.io/managed-by` label is better than my change.

**Comment by olemarkus** [comment]
With this branch, I could now do https://aws-controllers-k8s.github.io/community/docs/user-docs/install/
but instead of the IAM bit there, I just added this to the cluster spec:
```
    serviceAccountExternalPermissions:
    - aws:
        policyARNs:
        - arn:aws:iam::aws:policy/AmazonS3FullAccess
      name: ack-s3-controller
      namespace: ack-system
```

**Comment by olemarkus** [comment]
Can you clean up the commits in this PR? And move the template file?

Then I'll get a second opinion on the changes.

**Comment by olemarkus** [comment]
@h3poteto while we wait for upstream, perhaps you can add a PodDisruptionBudget and topologySpreadConstraint similar to most other addons?

**Comment by h3poteto** [comment]
@olemarkus OK, I will add it.

**Comment by olemarkus** [comment]
I added a commit with a few small changes. most notably not using a latest tag for the image.

I think I'd like to get this merged now even though it is not 100% ready upstream. that way I can work on some e2e tests and such to ensure this works as intended.

I can also work on extending the IRSA documentation a bit more as follow-up PRs

**Review by olemarkus — APPROVED** [review]

**Pull request merged**
